### ISSUE TITLE
`space-infix-ops`: add option for PEP8-style default parameters?

### ISSUE THREAD

**Issue by insin** [eslint/eslint#3587]
Would it be possible to add an option to `space-infix-ops` to allow or enforce assignment without spaces for ES6 default parameters?
### The use case for the rule - what is it trying to prevent or flag?

As of ESLint 1.2.0, `space-infix-ops` also applies to assignments. As such, developers who prefer to enforce spaces around assignments as per other ops but wish to omit them _only_ in default function parameters will now receive an unwanted warning or error.
### Whether the rule is trying to prevent an error or is purely stylistic

Purely stylistic, but intended to remove a false positive.
### Why you believe this rule is generic enough to be included

Default parameter value assignments are different from other types of variable assignments, in that they are part of a function's definition. This has different implications for how you read them as you scan through code. As such, developers may wish to apply different formatting rules to them according to their preferred style.

For an example from a community where this feature already exists, [Python's Style Guide (PEP8)](https://www.python.org/dev/peps/pep-0008/#other-recommendations) recommends spaces around all binary operators _except_ for when assigning default parameter values. It's likely that developers with Python experience who are using default parameters will already be using this style.
##

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by ilyavolodin** [comment]
Can you provide an example of the code that shouldn't error? And an error message you are currently getting for it? Also, do you have a name for this option in mind?

**Comment by insin** [comment]
With ESLint 1.3.1:

Config (via eslint-config-standard):

```
"space-infix-ops": 2,
```

Code (where the function is defined on line 14):

```js
function screen(state='search', action) {
  switch (action.type) {
    case LOGOUT:
      window.location.hash = ''
      return getDefaultSlug()
    case CHANGE_SCREEN:
      return action.slug
  }
  return state
}
```

Current error message:

```
app\ducks\screen.js
  14:22  error  Infix operators must be spaced  space-infix-ops
```

Naming things, that will required a sleep and/or a shower :smile: How about `compactDefaultParams` as an initial suggestion?

**Comment by BYK** [comment]
I was waiting for this issue since I touched `space-infix-ops` :) I like the name for `compactDefaultParams`. @eslint/eslint-team are we good on the name?

**Comment by BYK** [comment]
Well, one more question: should we enforce no spaces for these or simply ignore if they are not spaced? I feel like enforcing is better.

**Comment by insin** [comment]
Updated the OP with the info the contributing guide asks for.

I agree enforcing is better for the use case.

**Comment by BYK** [comment]
>I agree enforcing is better for the use case.

Now I realize this rule is designed to only enforce putting spaces but not lack of spaces. For this reason, I think the first step should be to simply ignore these in `FunctionDeclaration` and `FunctionExpression` nodes.

For enforcing no spaces, I think we need a breaking change for v2.0 that changes the first option to be `"always"` and `"never"` and moves the current options object to the second argument. Adding a `mode` argument as @btmills suggested for another rule to strictly enforce spaces (or lack of) or simply ignore certain cases would make it flexible enough. @eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
We'd probably want a new rule `space-ops` rather than continuing to hack this rule. We're already stretching the meaning of "infix" a bit too far.

In the short-term, what are we talking about here? Omitting default parameter spacing from this rule or enforcing no spaces?

**Comment by BYK** [comment]
>We'd probably want a new rule space-ops rather than continuing to hack this rule. We're already stretching the meaning of "infix" a bit too far.

:+1:

>In the short-term, what are we talking about here? Omitting default parameter spacing from this rule or enforcing no spaces?

I think adding the option to ignore this case is the way to go for short term. As mentioned before, this rule was all about enforcing spaces, not lack of, and it may be a bit too much for a single pass.

**Comment by nzakas** [comment]
Omitting this check is a breaking change, so it might be better to deprecate the existing rule and start work on a new one.

**Comment by BYK** [comment]
Okay, this works for me. So the name should be `space-ops` and it would allow enforcing putting spaces and not putting spaces with per-operator overrides, right?

**Comment by nzakas** [comment]
Per-operator and per-location

**Comment by bobbyrenwick** [comment]
@BYK what's the status of your new rule?

**Comment by fobdy** [comment]
+1

**Comment by nzakas** [comment]
@BYK are you still working on this?

**Comment by alberto** [comment]
It's been a long time without any movement.
Is this still a good fit for core?
Is anyone willing to implement this?

**Comment by BYK** [comment]
Sigh, I should start working on this as a new rule as agreed above: `space-ops`. I'm a bit swamped but I think this belongs to core.

**Comment by nzakas** [comment]
Adding "help wanted" and removing milestone. If we don't get any further movement in the next month, we should close this.

**Comment by insin** [comment]
> Per-operator and per-location

@nzakas Are there any recommendations/existing examples of what the config for this should look like to cleanly deal with the mixture of enforcing putting or omitting per operator and location?

**Comment by nzakas** [comment]
@insin no, we are looking for proposals.

**Comment by acontreras89** [comment]
Please do not close this. I know it's been almost a month, but I think this feature would be a really nice addition.
I've never worked on this repo before, but I would definitely give it a go if no one is already working on it.

**Comment by bobbyrenwick** [comment]
I've also been waiting with bated breath about this issue and would love to see it happen. Unfortunately I don't have the time to contribute at the moment but I'd support you all the way @acontreras89!

**Comment by ilyavolodin** [comment]
The issue is accepted and marked as "help wanted". That means that the team thinks that this is a good change, but we need somebody to implement it. However, as @nzakas said, we are first looking for complete proposal on what should change and with which option.

**Comment by onurtemizkan** [comment]
I've started working on this rule a last week but got distracted by another thing. I can finalize it in upcoming days if no one else is volunteering.

**Comment by onurtemizkan** [comment]
I've started working on this last week but got distracted by another thing. I can finalize it when the complete proposal is defined, if no one else is volunteering.

**Comment by acontreras89** [comment]
Go for it, @onurtemizkan :smiley:

**Comment by evpozdniakov** [comment]
Still not resolved since Aug 29, 2015?

**Comment by nzakas** [comment]
@evpozdniakov that's correct, if the issue is opened then it hasn't been resolved. The "help wanted" label indicates that it won't be done by the core team but a PR would be accepted.

**Comment by evpozdniakov** [comment]
I see. Lets say, I want to help. How do I start? What should I read before starting development, etc?
Thank you.

**Comment by nzakas** [comment]
@evpozdniakov this is the best place to start: http://eslint.org/docs/developer-guide/

**Comment by loganfuller** [comment]
@onurtemizkan are you still working on this? I'm planning on giving it a go this weekend.

**Comment by zertyuio** [comment]
Hi, any update about this issue ?

I am also interested by this new rule, it looks great !
I was too presumptuous this morning with the idea that I could maybe implement this new rule 'space-ops'.
I haven't even managed to update the existing one 'space-infix-ops' with an option, which sounds like the easiest way to implement it (but it's not obvious to understand how eslint works).

As I couldn't implement a clean solution, I have updated the file eslint/lib/rules/space-infix-ops.js and updated the file like this :

Original code:

```
return {
    AssignmentExpression: checkBinary,
    AssignmentPattern: checkBinary,
    BinaryExpression: checkBinary,
    LogicalExpression: checkBinary,
    ConditionalExpression: checkConditional,
    VariableDeclarator: checkVar
};
```

My code:

```
return {
    AssignmentExpression: checkBinary,
    // AssignmentPattern: checkBinary,
    BinaryExpression: checkBinary,
    LogicalExpression: checkBinary,
    ConditionalExpression: checkConditional,
    VariableDeclarator: checkVar
};
```

I'm aware how dirty this fix is + I'm not confident I can merge this update next time I will pull eslint, but it's a convenient working temporary fix. I'm sharing it in case other dirty coders are interested.

**Comment by machineghost** [comment]
Also wondering about the status of this issue.  If it's not too much trouble, it would probably be very helpful to outsiders if even a single sentence explaining the ticket's resolution was added when the ticket is closed.  Right now it's impossible to tell if this `space-ops` will be in the next version of ESLint, or whether the ESLint has simply decided it hates ES6 ;-)

**Comment by acontreras89** [comment]
The status of this issue is:

- It's a feature proposal regarding a new rule (labels _feature_, _rule_)
- It's been accepted (label _accepted_)
- Community help is wanted to get it implemented (label _help wanted_)

Moreover, this issue has not yet been closed.

Some people displayed interest in working on this issue. @onurtemizkan, @evpozdniakov, @loganfuller, has any of you made any progress at all?

**Comment by kaicataldo** [comment]
I think it's about time to close this if progress still hasn't been made.

**Comment by IanVS** [comment]
It looks like discussion about has moved over to https://github.com/eslint/eslint/issues/7489.

**Comment by kaicataldo** [comment]
Thanks! Closing this one in favor of that discussion then.
